src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -0,0 +1,757 @@
+
+	eopts := services.NewElectionOptions()
+
+	leaderOpts := leader.NewOptions().
+		SetServiceID(sid).
+		SetElectionOpts(eopts)
+
+	leaderService, err := leader.NewService(electionCluster.RandClient(), leaderOpts)
+	if err != nil {
+		return newAggregatorResult{}, err
+	}
+
+	electionManagerOpts := aggregator.NewElectionManagerOptions().
+		SetCampaignOptions(campaignOpts).
+		SetLeaderService(leaderService).
+		SetPlacementManager(placementManager).
+		SetFlushTimesManager(flushTimesManager)
+	electionManager := aggregator.NewElectionManager(electionManagerOpts)
+	aggregatorOpts = aggregatorOpts.SetElectionManager(electionManager)
+
+	// Set up flush manager.
+	flushManagerOpts := aggregator.NewFlushManagerOptions().
+		SetPlacementManager(placementManager).
+		SetFlushTimesManager(flushTimesManager).
+		SetElectionManager(electionManager)
+	flushManager := aggregator.NewFlushManager(flushManagerOpts)
+	aggregatorOpts = aggregatorOpts.SetFlushManager(flushManager)
+
+	flushWorkers := xsync.NewWorkerPool(storageFlushConcurrency)
+	flushWorkers.Init()
+	handler := newDownsamplerFlushHandler(o.Storage, sortedTagIteratorPool,
+		flushWorkers, instrumentOpts)
+	aggregatorOpts = aggregatorOpts.SetFlushHandler(handler)
+
+	return newAggregatorResult{
+		aggregator:              aggregator.NewAggregator(aggregatorOpts),
+		matcher:                 matcher,
+		tagEncoderPool:          tagEncoderPool,
+		tagDecoderPool:          tagDecoderPool,
+		encodedTagsIteratorPool: sortedTagIteratorPool,
+	}, nil
+}
+
+type downsamplerFlushHandler struct {
+	sync.RWMutex
+	storage                 storage.Storage
+	encodedTagsIteratorPool *encodedTagsIteratorPool
+	workerPool              xsync.WorkerPool
+	instrumentOpts          instrument.Options
+	metrics                 downsamplerFlushHandlerMetrics
+}
+
+type downsamplerFlushHandlerMetrics struct {
+	flushSuccess tally.Counter
+	flushErrors  tally.Counter
+}
+
+func newDownsamplerFlushHandlerMetrics(
+	scope tally.Scope,
+) downsamplerFlushHandlerMetrics {
+	return downsamplerFlushHandlerMetrics{
+		flushSuccess: scope.Counter("flush-success"),
+		flushErrors:  scope.Counter("flush-errors"),
+	}
+}
+
+func newDownsamplerFlushHandler(
+	storage storage.Storage,
+	encodedTagsIteratorPool *encodedTagsIteratorPool,
+	workerPool xsync.WorkerPool,
+	instrumentOpts instrument.Options,
+) handler.Handler {
+	scope := instrumentOpts.MetricsScope().SubScope("downsampler-flush-handler")
+	return &downsamplerFlushHandler{
+		storage:                 storage,
+		encodedTagsIteratorPool: encodedTagsIteratorPool,
+		workerPool:              workerPool,
+		instrumentOpts:          instrumentOpts,
+		metrics:                 newDownsamplerFlushHandlerMetrics(scope),
+	}
+}
+
+func (h *downsamplerFlushHandler) NewWriter(
+	scope tally.Scope,
+) (writer.Writer, error) {
+	return &downsamplerFlushHandlerWriter{
+		ctx:     context.Background(),
+		handler: h,
+	}, nil
+}
+
+func (h *downsamplerFlushHandler) Close() {
+}
```

Comment:
nit: can we get a comment here? Either `// noop` or `// TODO` depending on what's required